File: 64-minimum-path-sum/minimum-path-sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        int m=grid.length;
        int n=grid[0].length;
        int[][] dp=new int[m][n];
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                dp[i][j]=-1;
            }
        }
        return solve(grid,0,0,m,n,dp);
        
    }
    int solve(int[][] grid,int i,int j,int m,int n,int[][] dp){
             if(i>=m || j>=n){
                return Integer.MAX_VALUE;
            }
            if(dp[i][j]!=-1){
                return dp[i][j];
            }
            if(i==m-1 && j==n-1){
                return grid[i][j];
            }
          return dp[i][j]=grid[i][j]+Math.min(solve(grid,i+1,j,m,n,dp),solve(grid,i,j+1,m,n,dp));
           
          
    }
}

File: 64-minimum-path-sum/minimum-path-sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int n=grid.size();
        int m=grid[0].size();
        vector<int> prev(m,0);
        for(int i=0;i<n;i++){
            vector<int> curr(m,0);
            for(int j=0;j<m;j++){
                if(i==0 && j==0 ){
                    curr[j]=grid[i][j];
                    continue;
                }
                int up=0;
                int left=0;
                if(i>0) up=prev[j];
                else{
                    up=curr[j-1];
                }
                if(j>0) left=curr[j-1];
                else{
                    left=prev[j];
                }
                curr[j]=grid[i][j]+min(up,left);

            }prev=curr;
        }
        // for(int i=0;i<n;i++){
        //     for(int j=0;j<m;j++){
        //         cout << dp[i][j] << " ";
        //     }
        //     cout << '\n';
        // }
        return prev[m-1];
    }
};